lib/vdc/DerivedParticleDensity.cpp

```diff
@@ -0,0 +1,355 @@
+
+    int oxd = 1 + max[0] - min[0], oyd = 1 + max[1] - min[1], ozd = 1 + max[2] - min[2];
+    //    printf("Read(%s, %s, %s)\n", C(_derivedVarName), C(min), C(max));
+
+    // Copy subvolume
+    for (int z = 0; z < ozd; z++)
+        for (int y = 0; y < oyd; y++)
+            for (int x = 0; x < oxd; x++) region[z * oyd * oxd + y * oxd + x] = data[(min[2] + z) * yd * xd + (min[1] + y) * xd + (min[0] + x)];
+    //                region[z*oyd*oxd + y*oxd + x] = y % 3 ? 0 : 1;
+
+    //    for (size_t i = 0; i < nbins; i++)
+    //        region[i] = data[i];
+
+    delete[] data;
+    return 0;
+}
+
+void DerivedParticleDensity::compute(Grid *inGrid, float *output, int xd, int yd, int zd, size_t realNP) const
+{
+    DblArr3 minud, maxud;
+    inGrid->GetUserExtents(minud, maxud);
+    vec3 minu(minud[0], minud[1], minud[2]);
+    vec3 maxu(maxud[0], maxud[1], maxud[2]);
+    vec3 diffu = maxu - minu;
+
+    vec3    binDimsF(xd, yd, zd);
+    ivec3   binDimsI(xd, yd, zd);
+    size_t  nbins = xd * yd * zd;
+    size_t *bins = new size_t[nbins];
+    for (size_t i = 0; i < nbins; i++) bins[i] = 0;
+
+    auto it = inGrid->ConstCoordBegin();
+    auto end = inGrid->ConstCoordEnd();
+    for (size_t j = 0; it != end && j < realNP; ++it, ++j) {
+        vec3  c((*it)[0], (*it)[1], (*it)[2]);
+        vec3  t = (c - minu) / diffu;
+        ivec3 i = t * binDimsF;
+        i = glm::clamp(i, ivec3(0), binDimsI - ivec3(1, 1, 1));
+        bins[i.z * yd * xd + i.y * xd + i.x]++;
+    }
+
+    for (size_t i = 0; i < nbins; i++) output[i] = bins[i];
+
+    delete[] bins;
+}
+
+int DerivedParticleDensity::ReadRegionBlock(int fd, const std::vector<size_t> &min, const std::vector<size_t> &max, float *region) { return ReadRegion(fd, min, max, region); }
+
+int DerivedParticleDensity::GetDimLensAtLevel(int level, std::vector<size_t> &dims_at_level, std::vector<size_t> &bs_at_level) const
+{
+    dims_at_level.clear();
+    DC::Mesh mesh;
+    _dataMgr->GetMesh(_meshName, mesh);
+    DC::Dimension dim;
+    for (auto dimName : mesh.GetDimNames()) {
+        _dataMgr->GetDimension(dimName, dim, -1);
+        dims_at_level.push_back(dim.GetLength());
+    }
+
+    //    dims_at_level = {128, 128, 128};
+    bs_at_level = vector<size_t>(dims_at_level.size(), 1);
+    return 0;
+}
+
+bool DerivedParticleDensity::VariableExists(size_t ts, int reflevel, int lod) const { return true; }
+
+bool DerivedParticleDensity::GetBaseVarInfo(DC::BaseVar &var) const
+{
+    DC::DataVar dvar;
+    GetDataVarInfo(dvar);
+    var = dvar;
+    return true;
+}
+
+bool DerivedParticleDensity::GetDataVarInfo(DC::DataVar &dvar) const
+{
+    if (_dataVar.empty()) return false;
+
+    dvar.SetName(_derivedVarName);
+    dvar.SetMeshName(_meshName);
+    dvar.SetUnits("");
+    dvar.SetXType(DC::FLOAT);
+    dvar.SetWName("");
+    dvar.SetCRatios(vector<size_t>());
+    dvar.SetPeriodic(vector<bool>(3, false));
+
+    //    dvar.SetHasMissing(true);
+    //    dvar.SetMissingValue(-1);
```

Comment:
remove?

Reply:
I would like to leave this for future reference.